pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -17,106 +17,176 @@ limitations under the License.
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
 	})
 
-	c.informerStartFn(stopCh)
+	if !cache.WaitForNamedCacheSync("gkenetworkparamset", stopCh, c.gkeNetworkParamsInformer.HasSynced) {
+		return
+	}
+	for i := 0; i < numWorkers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
 
 	<-stopCh
-	klog.Infof("done gkenetworkparamset controller")
 }
 
-func (c *Controller) onGKENetworkParamSetAdded(params *networkv1alpha1.GKENetworkParamSet) error {
-	return c.updateGKENetworkParamSet(params)
+// worker pattern adapted from https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
 }
-func (c *Controller) onGKENetworkParamSetUpdated(oldParams, newParams *networkv1alpha1.GKENetworkParamSet) error {
-	return c.updateGKENetworkParamSet(newParams)
+
+func (c *Controller) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.updateGKENetworkParamSet(ctx, key.(string))
+	c.handleErr(err, key)
+	return true
 }
 
-func (c *Controller) updateGKENetworkParamSet(params *networkv1alpha1.GKENetworkParamSet) (rerr error) {
-	ctx := context.Background()
-	cidrs := []string{}
+// handleErr checks if an error happened and makes sure we will retry later.
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+		c.queue.Forget(key)
+		return
+	}
 
-	defer func() {
-		err := gkenetworkparamsetutil.UpdateGKENetworkParamSetStatus(ctx, c.networkClientset.NetworkingV1alpha1().GKENetworkParamSets(), params, cidrs)
-		if err != nil {
-			klog.Error(err, "failed to patch GKENetworkParamSetStatus")
-			rerr = multierr.Append(rerr, err)
-		}
-	}()
+	// This controller retries 5 times if something goes wrong. After that, it stops trying.
+	if c.queue.NumRequeues(key) < 5 {
+		klog.Warningf("Error while updating GKENetworkParamSet object, retrying %v: %v", key, err)
+
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+	// Report to an external entity that, even after several retries, we could not successfully process this key
+	utilruntime.HandleError(err)
+	klog.Errorf("Dropping GKENetworkParamSet %q out of the queue: %v", key, err)
+}
+
+func (c *Controller) updateGKENetworkParamSet(ctx context.Context, key string) error {
+	obj, exists, err := c.gkeNetworkParamsInformer.GetIndexer().GetByKey(key)
```

Comment:
this is prettier if you use the Lister directly, check how the endpoints controller is configured https://github.com/kubernetes/kubernetes/blob/6ce055d62d8608842db18901c948c6982c74a164/pkg/controller/endpoint/endpoints_controller.go#L122-L127

and how implement the reconcile loop 
https://github.com/kubernetes/kubernetes/blob/6ce055d62d8608842db18901c948c6982c74a164/pkg/controller/endpoint/endpoints_controller.go#L362-L366

Reply:
forget the comment, as you wish, I didn't notice you use the indexer for the `exist` output